### PR TITLE
Improved handling of 100 Continue

### LINKED ISSUE #11841: Cannot invoke "org.eclipse.jetty.client.transport.HttpExchange.getRequest()" because "exchange" is null
**Jetty version(s)**
12.0.9

**Jetty Environment**
ee9

Java
```
openjdk version "17.0.11" 2024-04-16
OpenJDK Runtime Environment Temurin-17.0.11+9 (build 17.0.11+9)
OpenJDK 64-Bit Server VM Temurin-17.0.11+9 (build 17.0.11+9, mixed mode, sharing)
```
Maven
```
Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)
Maven home: C:\java\apache-maven-3.9.6
Java version: 17.0.11, vendor: Eclipse Adoptium, runtime: C:\Program Files\Eclipse Adoptium\jdk-17.0.11.9-hotspot
Default locale: en_US, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```

**OS type/version**
```
Microsoft Windows [Version 10.0.19045.4291]
```

**Description**

```
java.lang.NullPointerException: Cannot invoke "org.eclipse.jetty.client.transport.HttpExchange.getRequest()" because "exchange" is null
	at org.eclipse.jetty.client.transport.HttpSender.abortRequest(HttpSender.java:237)
	at org.eclipse.jetty.client.transport.HttpSender.internalAbort(HttpSender.java:389)
	at org.eclipse.jetty.client.transport.HttpSender$ContentSender.onCompleteFailure(HttpSender.java:605)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:335)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:231)
	at org.eclipse.jetty.client.transport.HttpSender.send(HttpSender.java:85)
	at org.eclipse.jetty.client.transport.internal.HttpChannelOverHTTP.send(HttpChannelOverHTTP.java:86)
	at org.eclipse.jetty.client.transport.HttpChannel.send(HttpChannel.java:142)
	at org.eclipse.jetty.client.transport.HttpConnection.send(HttpConnection.java:112)
	at org.eclipse.jetty.client.transport.internal.HttpConnectionOverHTTP$Delegate.send(HttpConnectionOverHTTP.java:330)
	at org.eclipse.jetty.client.transport.internal.HttpConnectionOverHTTP.send(HttpConnectionOverHTTP.java:159)
	at org.eclipse.jetty.client.transport.HttpDestination.send(HttpDestination.java:444)
	at org.eclipse.jetty.client.transport.HttpDestination.process(HttpDestination.java:420)
	at org.eclipse.jetty.client.transport.HttpDestination.process(HttpDestination.java:375)
	at org.eclipse.jetty.client.transport.HttpDestination.send(HttpDestination.java:358)
	at org.eclipse.jetty.client.transport.HttpDestination.send(HttpDestination.java:352)
	at org.eclipse.jetty.client.transport.HttpDestination.send(HttpDestination.java:329)
	at org.eclipse.jetty.client.transport.HttpDestination.send(HttpDestination.java:308)
	at org.eclipse.jetty.client.transport.HttpRequest.sendAsync(HttpRequest.java:751)
	at org.eclipse.jetty.client.transport.HttpDestination.send(HttpDestination.java:303)
	at org.eclipse.jetty.client.transport.HttpRequest.send(HttpRequest.java:744)
	at org.eclipse.jetty.client.CompletableResponseListener.send(CompletableResponseListener.java:79)
	at org.eclipse.jetty.client.transport.HttpRequest.send(HttpRequest.java:707)
	at _unit test stack frame_
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at _test stack frame_
```

Perhaps, coincidentally, I just saw a similar stack trace with an another test; note that our custom servlet subclasses Jetty's AsyncMiddleManServlet, so I am not sure if the stack traces below are from the client side since the test uses the Jetty client or the server side since the AsyncMiddleManServlet also uses the Jetty client.

```
2024-05-24 15:39:45,009 WARN  [HttpClient@3c1e23ff-24][o.e.j.i.SelectorManager] Exception while notifying connection HttpConnectionOverHTTP@7de18c7::SocketChannelEndPoint@3e19ce4[{l=/127.0.0.1:62969,r=localhost/127.0.0.1:59815,OPEN,fill=FI,flush=-,to=0/30000}{io=1/1,kio=1,kro=1}]->[HttpConnectionOverHTTP@7de18c7(l:/127.0.0.1:62969 <-> r:localhost/127.0.0.1:59815,closed=false)=>HttpChannelOverHTTP@46e312fc(exchange=HttpExchange@e8e3{req=HttpRequest[PATCH /api/abc00.txt HTTP/1.1]@77efd004[TERMINATED/null] res=HttpResponse[HTTP/1.1 100 Continue]@78e674a1[PENDING/null]})[send=HttpSenderOverHTTP@26ccedb3(req=FAILURE,failure=java.lang.NullPointerException: Cannot invoke "org.eclipse.jetty.client.transport.HttpExchange.getRequest()" because "this.exchange" is null)[HttpGenerator@218c4e79{s=START}],recv=HttpReceiverOverHTTP@286ac5bc(ex=HttpExchange@e8e3{req=HttpRequest[PATCH /api/abc00.txt HTTP/1.1]@77efd004[TERMINATED/null] res=HttpResponse[HTTP/1.1 100 Continue]@78e674a1[PENDING/null]},rsp=IDLE,failure=null)[HttpParser{s=START,0 of -1}]]]
java.lang.NullPointerException: Cannot invoke "org.eclipse.jetty.client.transport.HttpExchange.getRequest()" because "exchange" is null
	at org.eclipse.jetty.client.transport.HttpSender.abortRequest(HttpSender.java:237) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.HttpSender.internalAbort(HttpSender.java:389) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.HttpSender$ContentSender.onCompleteFailure(HttpSender.java:605) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:335) ~[jetty-util-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:231) ~[jetty-util-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.HttpSender.send(HttpSender.java:85) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.internal.HttpChannelOverHTTP.send(HttpChannelOverHTTP.java:86) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.HttpChannel.send(HttpChannel.java:142) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.HttpConnection.send(HttpConnection.java:112) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.internal.HttpConnectionOverHTTP$Delegate.send(HttpConnectionOverHTTP.java:330) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.internal.HttpConnectionOverHTTP.send(HttpConnectionOverHTTP.java:159) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.HttpDestination.send(HttpDestination.java:444) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.HttpDestination.process(HttpDestination.java:420) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.HttpDestination.process(HttpDestination.java:375) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.HttpDestination.send(HttpDestination.java:358) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.HttpDestination.succeeded(HttpDestination.java:291) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.AbstractConnectionPool.proceed(AbstractConnectionPool.java:315) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.AbstractConnectionPool$FutureConnection.succeeded(AbstractConnectionPool.java:615) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.AbstractConnectionPool$FutureConnection.succeeded(AbstractConnectionPool.java:593) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.util.Promise$Wrapper.succeeded(Promise.java:195) ~[jetty-util-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.internal.HttpConnectionOverHTTP.onOpen(HttpConnectionOverHTTP.java:167) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.io.SelectorManager.connectionOpened(SelectorManager.java:325) [jetty-io-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.io.ClientConnector$ClientSelectorManager.connectionOpened(ClientConnector.java:616) [jetty-io-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.io.ManagedSelector.createEndPoint(ManagedSelector.java:401) [jetty-io-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.io.ManagedSelector$CreateEndPoint.run(ManagedSelector.java:1056) [jetty-io-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979) [jetty-util-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209) [jetty-util-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164) [jetty-util-12.0.9.jar:12.0.9]
	at java.base/java.lang.Thread.run(Thread.java:840) [?:?]
2024-05-24 15:39:45,027 WARN  [HttpClient@3c1e23ff-18][o.e.j.i.SelectorManager] Exception while notifying connection HttpConnectionOverHTTP@29cf8d83::SocketChannelEndPoint@2d3f260[{l=/[0:0:0:0:0:0:0:1]:62970,r=localhost/[0:0:0:0:0:0:0:1]:59815,OPEN,fill=FI,flush=-,to=2/30000}{io=1/1,kio=1,kro=8}]->[HttpConnectionOverHTTP@29cf8d83(l:/[0:0:0:0:0:0:0:1]:62970 <-> r:localhost/[0:0:0:0:0:0:0:1]:59815,closed=false)=>HttpChannelOverHTTP@565aa3bb(exchange=null)[send=HttpSenderOverHTTP@149f6dee(req=QUEUED,failure=null)[HttpGenerator@41e9496b{s=START}],recv=HttpReceiverOverHTTP@5cb4f3c4(ex=null,rsp=IDLE,failure=null)[HttpParser{s=START,0 of -1}]]]
java.lang.IllegalStateException: Entry already enabled ConcurrentEntry@701b6587{terminated=false,multiplex=0,pooled=HttpConnectionOverHTTP@7de18c7::SocketChannelEndPoint@3e19ce4[{l=/127.0.0.1:62969,r=localhost/127.0.0.1:59815,OPEN,fill=FI,flush=-,to=1/30000}{io=1/1,kio=1,kro=1}]->[HttpConnectionOverHTTP@7de18c7(l:/127.0.0.1:62969 <-> r:localhost/127.0.0.1:59815,closed=false)=>HttpChannelOverHTTP@46e312fc(exchange=null)[send=HttpSenderOverHTTP@26ccedb3(req=FAILURE,failure=java.lang.NullPointerException: Cannot invoke "org.eclipse.jetty.client.transport.HttpExchange.getRequest()" because "this.exchange" is null)[HttpGenerator@218c4e79{s=START}],recv=HttpReceiverOverHTTP@286ac5bc(ex=null,rsp=IDLE,failure=null)[HttpParser{s=START,0 of -1}]]]} for ConcurrentPool@1bf9f49d[strategy=FIRST,inUse=0,size=1,max=64,leaked=0,terminated=false]
	at org.eclipse.jetty.util.ConcurrentPool$ConcurrentEntry.enable(ConcurrentPool.java:489) ~[jetty-util-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.AbstractConnectionPool$FutureConnection.succeeded(AbstractConnectionPool.java:612) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.AbstractConnectionPool$FutureConnection.succeeded(AbstractConnectionPool.java:593) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.util.Promise$Wrapper.succeeded(Promise.java:195) ~[jetty-util-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.client.transport.internal.HttpConnectionOverHTTP.onOpen(HttpConnectionOverHTTP.java:167) ~[jetty-client-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.io.SelectorManager.connectionOpened(SelectorManager.java:325) [jetty-io-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.io.ClientConnector$ClientSelectorManager.connectionOpened(ClientConnector.java:616) [jetty-io-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.io.ManagedSelector.createEndPoint(ManagedSelector.java:401) [jetty-io-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.io.ManagedSelector$CreateEndPoint.run(ManagedSelector.java:1056) [jetty-io-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:979) [jetty-util-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1209) [jetty-util-12.0.9.jar:12.0.9]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1164) [jetty-util-12.0.9.jar:12.0.9]
	at java.base/java.lang.Thread.run(Thread.java:840) [?:?]
```
Note that the test hangs, perhaps due to the second stack trace in the block above.

**How to reproduce?**

The reproducer is in PR: https://github.com/jetty/jetty.project/pull/11891, but it is "random", sometimes it happens, sometimes is does not. Using Microsoft's SysInternal's [CPU Stres](https://learn.microsoft.com/en-us/sysinternals/downloads/cpustres) tool helps recreate the issue.

This stack trace is from a test case using a Jetty 12 HTTP Client to a Jetty 12 Server using a custom AsyncMiddleManServlet subclass. I can't share any of it unfortunately.

I am wondering if it is possible that there is some asynchronous bad luck involved here.

### PR DESCRIPTION
**Pull request by sbordet**
* Now `HttpClient` removed the `Expect` header if there is no request content.
* Changed AbstractProxyServlet and ProxyHandler check for request content: now the Content-Type header is not taken into consideration.
* Now the server avoids sending the 100 Continue response if there is no request content.

### CODE REVIEW

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java
Comment (gregw) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:1410]: This is not a sufficient test to check if a request has content.  Need to also check content-encoding at least.  See https://datatracker.ietf.org/doc/html/rfc9112#name-message-body-length.   Like `hasContent` in `AbstractProxyServlet`

Also would it not be better to not set `_expects100Continue` to `true` if there is no content?

**Review by gregw — DISMISSED** [review]

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:1410]: @sbordet Perhaps to the check for a body in org.eclipse.jetty.http.MetaData.Request#is100ContinueExpected or org.eclipse.jetty.server.internal.HttpChannelState#onRequest?

**Comment by joakime** [comment]
Is the work here going to help the old 100-Continue issue with AsyncMiddleManServlet?

* #9758

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:1410]: Oh I think this is OK now because it is the inverse!

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpConnection.java
Comment (gregw) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpConnection.java]:
```suggestion
        if (StringUtil.isBlank(path))
```

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpConnection.java
Comment (gregw) [inline jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpConnection.java:213]: technically you should just remove the 100-continue value from the Expect header, but as there are currently no other defined values, this is probably OK.

File: jetty-ee9/jetty-ee9-proxy/src/main/java/org/eclipse/jetty/ee9/proxy/AbstractProxyServlet.java
Comment (gregw) [inline jetty-ee9/jetty-ee9-proxy/src/main/java/org/eclipse/jetty/ee9/proxy/AbstractProxyServlet.java]: A little javadoc explaining the return would be good, specifically that null makes it optional

**Review by gregw — DISMISSED** [review]
OK, but a couple of niggles.

**Comment by sbordet** [comment]
@gregw I fixed the niggles in #12111. Not a big fan of `StringUtil` when I can use JDK methods.

@joakime the work here does not address #9758, which I will look at separately.

**Review by firoj0 — COMMENTED** [review]
Git

**Review by firoj0 — APPROVED** [review]

**Comment by gregw** [comment]
> @gregw I fixed the niggles in #12111. Not a big fan of `StringUtil` when I can use JDK methods.
> 

In this case, the StringUtil method avoids an allocation and is much clearer to read

**Review by lorban — DISMISSED** [review]

**Comment by garydgregory** [comment]
Hi All,

Would you please set expectations as to next steps? Are we waiting on a review from @gregw ?

TY!

File: jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpConnection.java
Comment (sbordet) [reply jetty-core/jetty-client/src/main/java/org/eclipse/jetty/client/transport/HttpConnection.java:213]: Correct. And we don't have an easy API in `HttpFields` to remove a single value.

File: jetty-ee10/jetty-ee10-proxy/src/test/java/org/eclipse/jetty/ee10/proxy/ProxyServletTest.java
Comment (lorban) [inline jetty-ee10/jetty-ee10-proxy/src/test/java/org/eclipse/jetty/ee10/proxy/ProxyServletTest.java:1]: Shouldn't these tests be ported to ee9 too?

**Review by lorban — CHANGES_REQUESTED** [review]

**Comment by sbordet** [comment]
@garydgregory this PR is almost ready to be merged in, we won't wait for @gregw (currently in vacation).

It will be in the next Jetty 12 release, due end of month.

File: jetty-ee10/jetty-ee10-proxy/src/test/java/org/eclipse/jetty/ee10/proxy/ProxyServletTest.java
Comment (sbordet) [reply jetty-ee10/jetty-ee10-proxy/src/test/java/org/eclipse/jetty/ee10/proxy/ProxyServletTest.java:1]: Done.

**Review by lorban — APPROVED** [review]

**Pull request merged**
